Initial Support for Egis ET171

jackylee-go opened the pull request:
This PR adds initial support for Egis ET171.
This is a SOC based on Andes AE350 and integrates USB and other interfaces.
Other interfaces driver will be added in later PRs.

nordicjm commented on boards/egis/egis_et171/Kconfig.defconfig:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2025 Egis Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD
+	default "egis_et171" if BOARD_EGIS_ET171
```
hwmv1 stuff to go

nordicjm commented on boards/egis/egis_et171/Kconfig.defconfig:
```diff
@@ -0,0 +1,8 @@
+
+config BOARD
+	default "egis_et171" if BOARD_EGIS_ET171
+
+config FLASH_LOAD_SIZE
+	default 0x100000
```
why is this needed?

nordicjm commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,78 @@
+CONFIG_XIP=n
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+CONFIG_FLASH_BASE_ADDRESS=0x80000000
```
??

nordicjm replied:
use dts

nordicjm commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,78 @@
+# GPIO driver options
+CONFIG_GPIO=y
+
+# Syscon driver options
+CONFIG_SYSCON=y
+
+# hwinfo driver options
+CONFIG_HWINFO=y
+CONFIG_HWINFO_ANDES=y
+
+# RV64 options
+#CONFIG_RV64I_CPU=y
+
+# FPU (single precision) options
+#CONFIG_FPU=y
+#CONFIG_FPU_SHARING=y
+#CONFIG_FLOAT_HARD=y
+#CONFIG_SINGLE_PRECISION_FPU=y
+
+# FPU (double precision) options
+#CONFIG_FPU=y
+#CONFIG_FPU_SHARING=y
+#CONFIG_FLOAT_HARD=y
+#CONFIG_DOUBLE_PRECISION_FPU=y
+
+# SRAM size KB
+CONFIG_SRAM_SIZE=384
+
+# PMP options
+#CONFIG_PMP_GRANULARITY=8
+
+# SMP options
+#CONFIG_SMP=y
+#CONFIG_MP_MAX_NUM_CPUS=$(cores_num)
+
```
why is any of this, or anything below here?

jimmyzhe replied:
@jackylee-go, I am updating adp_xc7k_ae350_defconfig in #85927, you may refer to it.

nordicjm commented on dts/riscv/egis/egis_et171.dtsi:
```diff
@@ -0,0 +1,149 @@
+			clock-frequency = <66000000>;
+			i-cache-line-size = <32>;
+			d-cache-line-size = <32>;
+			CPU0_intc: interrupt-controller {
```
newline on line 29 and follow dts naming requirements https://docs.zephyrproject.org/latest/contribute/style/devicetree.html also note the linux ones

nordicjm commented on soc/andestech/ae350/et171_init.c:
```diff
@@ -0,0 +1,23 @@
+
+	return 0;
+}
+SYS_INIT(et171_soc_init, EARLY, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
use soc hook

nordicjm commented on soc/andestech/ae350/Kconfig.defconfig.et171:
```diff
@@ -0,0 +1,32 @@
+config SYS_CLOCK_TICKS_PER_SEC
+	default 100 if (!ICACHE || XIP)
+
+config MAIN_STACK_SIZE
+	default 2048
+
+config IDLE_STACK_SIZE
```
why?

nordicjm commented on soc/andestech/ae350/Kconfig.defconfig:
```diff
@@ -7,6 +7,7 @@ if SOC_SERIES_ANDES_AE350
 rsource "Kconfig.defconfig.ae*"
+rsource "Kconfig.defconfig.et171"
 
 config SYS_CLOCK_HW_CYCLES_PER_SEC
 	default 60000000
```
update this is get value from dts

jimmyzhe commented:
Thanks for using Andes feature. But I think it should be `soc/egis/et171` rather than integrating it into `soc/andestech/ae350`.

Although the ET171 SoC licenses Andes IP, it is not an Andes product. Using `soc/egis` allows easier extension if Egis adds more SoCs in the future..

For Andes peripheral IPs (e.g. `andestech,atcgpio100`), you can reuse existing drivers and device tree.
For Andes ISA extensions (CSR, instruction and core feature .etc), since `arch/riscv` currently does not support vendor-specific extensions (e.g. Xandes, XThead),  you may refer to [`soc/telink/tlsr
/tlsr951x`](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/telink/tlsr/tlsr951x), copy Andes-related extensions, and implement ET171-specific initialization code accordingly.

And I think you not need all Kconfig options/features from [soc/andestech/ae350/Kconfig](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/andestech/ae350/Kconfig) because ET171 should be a fixed SoC, while AE350 is a configurable platform with multiple CPU core options (e.g. AE350-N25, AE350-AX45) and we nedd these options for configuration.

jackylee-go commented:
> Thanks for using Andes feature. But I think it should be `soc/egis/et171` rather than integrating it into `soc/andestech/ae350`.
> 
> Although the ET171 SoC licenses Andes IP, it is not an Andes product. Using `soc/egis` allows easier extension if Egis adds more SoCs in the future..
> 
> For Andes peripheral IPs (e.g. `andestech,atcgpio100`), you can reuse existing drivers and device tree. For Andes ISA extensions (CSR, instruction and core feature .etc), since `arch/riscv` currently does not support vendor-specific extensions (e.g. Xandes, XThead), you may refer to [`soc/telink/tlsr/tlsr951x`](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/telink/tlsr/tlsr951x), copy Andes related extensions, and implement ET171-specific initialization code accordingly.
> 
> And I think you don't need all Kconfig options/features from [soc/andestech/ae350/Kconfig](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/andestech/ae350/Kconfig) because ET171 should be a fixed SoC, while AE350 is a configurable platform with multiple CPU core options (e.g. AE350-N25, AE350-AX45) and we need these options for configuration.

Yes, we think so too.
But to ensure complete functionality and future maintainability.
`soc/egis/et171` will duplicate almost all of `soc/andestech/ae350` content.
This will result in a very high duplication rate.
If this does not cause PR restrictions, we are happy to do so.

jimmyzhe commented:
I think it is ok to duplicate some code (as done for [tlsr951x](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/telink/tlsr/tlsr951x)) since we currently don't have a proper RISC-V vendor-extension ISA framework. I am more concern about the SoC  [hardware model](https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#soc), as the vendor of ET171 is Egis.

By the way, some `soc/andestech/ae350/Kconfig` options are not suitable for ET171.
For examples:
1. ET171 should not need Base ISA options
 https://github.com/zephyrproject-rtos/zephyr/blob/af4e875ea5b79ec2bf0b046939e5d802cc40fa8c/soc/andestech/ae350/Kconfig#L25
2. Similarly, FPU options are not necessary https://github.com/zephyrproject-rtos/zephyr/blob/af4e875ea5b79ec2bf0b046939e5d802cc40fa8c/soc/andestech/ae350/Kconfig#L50
3. If ET171 support PPMA, just initialize nocache memory directly and no need for  `SOC_ANDES_V5_PMA`
https://github.com/zephyrproject-rtos/zephyr/blob/af4e875ea5b79ec2bf0b046939e5d802cc40fa8c/soc/andestech/ae350/Kconfig#L92

4. ... (Other features not supported by ET171)

niedzwiecki-dawid commented on soc/egis/et171/start.S:
```diff
@@ -0,0 +1,87 @@
+	j header_end
+	nop /* @TODO: use to align. ".align" 2 doesn't work */
+	.word 0x53494745 /* "EGIS" */
+	.word 0xaaaaaaaa /* version */
```
Can you use a config for that?

niedzwiecki-dawid commented on soc/egis/et171/start.S:
```diff
@@ -0,0 +1,87 @@
+	nop /* @TODO: use to align. ".align" 2 doesn't work */
+	.word 0x53494745 /* "EGIS" */
+	.word 0xaaaaaaaa /* version */
+	.word __rom_region_start /* Entry address */
```
Can you also use configs for these values?

rruuaanng commented on dts/riscv/egis/egis_et171.dtsi:
```diff
@@ -0,0 +1,161 @@
+			compatible = "andestech,andescore-v5", "riscv";
+			device_type = "cpu";
+			reg = <0>;
+			status = "okay";
```
It's redundant.

nordicjm commented on boards/egis/egis_et171/Kconfig.egis_et171:
```diff
@@ -0,0 +1,11 @@
+config BOARD_EGIS_ET171
+	select SOC_EGIS_ET171
+
+config FIRMWARE_VERSION
```
```suggestion
config BOARD_FIRMWARE_VERSION
```

nordicjm commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,34 @@
+# EXEC.IT options
+#CONFIG_SOC_ANDES_V5_EXECIT=y
+
+# PLMT clock cycle presecond
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=18000000
```
not allowed, get value from dts node

nordicjm commented on soc/egis/soc.yml:
```diff
@@ -0,0 +1,6 @@
+family:
+- name: andes_v5
```
can't have the same family for a complete different device as an existing family

jimmyzhe commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_GPIO=y
+
+# Syscon driver options
+CONFIG_SYSCON=y
```
No need for this, it will be removed for `adp_xc7k_ae350` since it is not used during platform initialization.

jimmyzhe commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,34 @@
+
+# hwinfo driver options
+CONFIG_HWINFO=y
+CONFIG_HWINFO_ANDES=y
```
Same reason as CONFIG_SYSCON.

jimmyzhe commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_FPU=y
+#CONFIG_FPU_SHARING=y
+CONFIG_FLOAT_HARD=y
+CONFIG_SINGLE_PRECISION_FPU=y
```
The FPU feature should be selected at the SoC level.

niedzwiecki-dawid replied:
Makes sense

jimmyzhe commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,34 @@
+CONFIG_SINGLE_PRECISION_FPU=y
+
+# PMP granularity options
+CONFIG_PMP_GRANULARITY=8
```
Same reason as CONFIG_FPU.

jimmyzhe commented on dts/riscv/egis/egis_et171.dtsi:
```diff
@@ -0,0 +1,160 @@
+	soc {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "andestech,ae350";
```
Should be egis,et171.

jimmyzhe reviewed (commented):
ET171 is a product from Egis and should define Egis SoC family, series, and SoC layer, instead of selecting SOC_FAMILY_ANDES_V5.

As a fixed feature SoC, ET171 should not simply reuse the AE350 Kconfig interface.
Could you provide more details about ET171's core features?
Feature like FPU, PMA, and PMP granularity ... should be defined at the SoC level, not exposed in egis_et171_defconfig or menuconfig.

Aside from ET171's SoC definition, I believe what you really need is support for Andes extended CSRs.
I'm still figuring out how to add Custom CSR interface based on the [RISC-V Privileged Architecture Spec](https://drive.google.com/file/d/17GeetSnT5wW3xNuAHI95-SI1gPGd5sJ_/view) (2.1 CSR Address Mapping Conventions) to centralize support for Andes CSR feature.

Before that, could you refer to [Telink tlsr951x](https://github.com/zephyrproject-rtos/zephyr/blob/eb06ff8d3af371cfbdba5093dbc2c7ed69fffe66/soc/telink/tlsr/tlsr951x/Kconfig#L24) did and change the CSR and Kconfig prefixes from Andes to Egis.

niedzwiecki-dawid commented on soc/egis/et171/start.S:
```diff
@@ -0,0 +1,75 @@
+	j header_end
+	nop /* @TODO: use to align. ".align" 2 doesn't work */
+	.word 0x53494745 /* "EGIS" */
+	.word CONFIG_FIRMWARE_VERSION /* version */
```
This should be defined in SOC configs, not board configs.

jackylee-go commented on dts/riscv/egis/egis_et171.dtsi:
```diff
@@ -0,0 +1,159 @@
+		};
+
+		syscon: syscon@f0100000 {
+			compatible = "syscon", "egis,et171-aosum";
```
Which word? "aosum" means "Always On SMU".

niedzwiecki-dawid replied:
aosum -> aosmu

nordicjm commented on boards/egis/egis_et171/Kconfig.defconfig:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+DT_CHOSEN_Z_FLASH := zephyr,flash
+
```
split into multiple commits, one for soc, one for board

nordicjm commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,19 @@
+
+# GPIO driver options
+CONFIG_GPIO=y
+
+# FPU (single precision) options
+#CONFIG_FPU=y
+#CONFIG_FLOAT_HARD=y
+
+# HW DSP options
+#CONFIG_SOC_ANDES_V5_HWDSP=y
+
+# EXEC.IT options
+#CONFIG_SOC_ANDES_V5_EXECIT=y
```
```suggestion
```

nordicjm commented on soc/egis/et171/Kconfig:
```diff
@@ -0,0 +1,30 @@
+	select HAS_FLASH_LOAD_OFFSET
+	select SOC_EARLY_INIT_HOOK
+
+config SOC_FIRMWARE_VERSION
+	hex "Firmware version"
+	default 0xAAAAAAAA
+	help
+	  Firmware version info of bin header
```
this needs renaming to something specific of the soc

nordicjm commented on soc/egis/et171/soc_context.h:
```diff
@@ -0,0 +1,44 @@
+#ifdef CONFIG_RISCV_SOC_CONTEXT_SAVE
+
+/* Andes V5 specific registers. */
+#if defined(CONFIG_SOC_ANDES_V5_PFT) && defined(CONFIG_SOC_ANDES_V5_HWDSP)
```
I don't see why this file is using andes soc Kconfigs

nordicjm commented on soc/egis/et171/soc_offsets.h:
```diff
@@ -0,0 +1,34 @@
+#ifdef CONFIG_RISCV_SOC_OFFSETS
+
+/* Andes V5 specific registers. */
+#if defined(CONFIG_SOC_ANDES_V5_PFT) && defined(CONFIG_SOC_ANDES_V5_HWDSP)
```
as above

nordicjm commented on soc/egis/Kconfig:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_FAMILY_ANDES_V5
```
no hijacking of Kconfigs, this Kconfig has nothing to do with this soc, remove instances of it below

jimmyzhe reviewed (commented):
I would still suggest to duplicate Andes CSR code into soc/egis/et171 as [soc/telink/tlsr/tlsr951x](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/telink/tlsr/tlsr951x) did.
Egis should define `SOC_FAMILY_EGIS...`, `SOC_SERIES_EGIS...` instead of using `SOC_FAMILY_ANDES_V5`.

If #93697 is accepted, the soc/egis/et171 can apply the new custom csr framework.
If this PR is merged before #93697 (as telink did), I can help to update ET171 with the new custom csr framework.

`SOC_FAMILY_ANDES_V5` should not be used in soc/egis, because Andestech is not the vendor of ET171.

jackylee-go commented:
> I would still suggest to duplicate Andes CSR code into soc/egis/et171 as [soc/telink/tlsr/tlsr951x](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/telink/tlsr/tlsr951x) did. Egis should define `SOC_FAMILY_EGIS...`, `SOC_SERIES_EGIS...` instead of using `SOC_FAMILY_ANDES_V5`.
> 
> If #93697 is accepted, the soc/egis/et171 can apply the new custom csr framework. If this PR is merged before #93697 (as telink did), I can help to update ET171 with the new custom csr framework.
> 
> `SOC_FAMILY_ANDES_V5` should not be used in soc/egis, because Andestech is not the vendor of ET171.

Since the ET171 is a custom AE350-based MCU, I'd like to reuse existing AE350 drivers, such as andes_cache. However, its Kconfig.andes requires the SOC_FAMILY_ANDES_V5 declaration to use it correctly.
[drivers/cache/Kconfig.andes](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/cache/Kconfig.andes#L5-L11) 
```c
config CACHE_ANDES
bool "ANDES external cache driver"
	default y
	depends on SOC_FAMILY_ANDES_V5
	select CACHE_HAS_DRIVER
	help
	  This option enables the CACHE driver for ANDES V5 series SOC.
```
[drivers/cache/CMakeLists.txt](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/cache/CMakeLists.txt#L9C1-L9C64)
```c
zephyr_library_sources_ifdef(CONFIG_CACHE_ANDES 	cache_andes.c)

```
Do you have any suggestions for reusing this driver for compatible MCUs?
Or would it be more appropriate to rewrite a cache driver specifically for the ET171 (even if the code is nearly duplicated)?

jackylee-go commented on soc/egis/et171/Kconfig:
```diff
@@ -0,0 +1,30 @@
+	select CPU_HAS_FPU
+	select CPU_HAS_DCACHE
+	select CPU_HAS_ICACHE
+	select CACHE_MANAGEMENT if DCACHE
```
No, that's because the next driver I would like to PR after currnet PR will need cache control (if dcash is enabled).\
So, I need to select it when enabling DCACHE.

jimmyzhe replied:
I suggest  to add `select CACHE_MANAGEMENT if DCACHE` in that driver's Kconfig, so that all SoC using that driver have this dependency, not just ET171.

jackylee-go commented on soc/egis/et171/Kconfig.defconfig:
```diff
@@ -0,0 +1,55 @@
+	default 8
+
+config PMP_NO_NA4
+	default n
```
Thanks for the reminder.
It should be "y", I mixed up the reverse logic.
```
config PMP_NO_NA4
   default y
```

jimmyzhe commented:
> Do you have any suggestions for reusing this driver for compatible MCUs? Or would it be more appropriate to rewrite a cache driver specifically for the ET171 (even if the code is nearly duplicated)?

We can modify [drivers/cache/Kconfig.andes](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/cache/Kconfig.andes#L5-L11) in this PR.
```
config CACHE_ANDES
bool "ANDES external cache driver"
	default y
-	depends on SOC_FAMILY_ANDES_V5
+	depends on DT_HAS_ANDESTECH_ANDESCORE_V5_ENABLED
	select CACHE_HAS_DRIVER
	help
	  This option enables the CACHE driver for ANDES V5 series SOC.
```

jackylee-go commented on soc/egis/et171/Kconfig.defconfig.et171:
```diff
@@ -0,0 +1,21 @@
+	default $(dt_node_int_prop_int,/cpus/cpu@0,d-cache-line-size)
+
+config ICACHE_LINE_SIZE
+	default $(dt_node_int_prop_int,/cpus/cpu@0,i-cache-line-size)
```
Since [USB](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/usb/udc_buf.h#L26) devices require `CONFIG_DCACHE_LINE_SIZE` and `CONFIG_ICACHE_LINE_SIZE`, declaring it in Kconfig.defconfig.et171 should solve this problem better.

jimmyzhe replied:
After #95324 is merged, default value of `ICACHE_LINE_SIZE`, `DCACHE_LINE_SIZE` match ET171 requirement. 
https://github.com/zephyrproject-rtos/zephyr/blob/c7e60346a03eec8604c46c3b8bc98d22b6e8216f/arch/Kconfig#L1152-L1153

jackylee-go replied:
Thanks for your mention, I will remove it after rebasing.

jackylee-go commented on soc/egis/Kconfig.soc:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2025 Egis Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FAMILY_ANDES_V5
```
This is an invalid configuration and we wouldn't select it.
I will remove it.

jackylee-go commented on soc/egis/et171/start.S:
```diff
@@ -0,0 +1,75 @@
+	li t0, (1 << 19) | (1 << 13) | (1 << 10) | (1 << 1)
+	csrs NDS_MCACHE_CTL, t0
+
+	/* Check if CPU support CM or not. */
```
However, we still need cache management to flush and write back cache data.
As your [mentioned](https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2314832598) before, we will only select `CACHE_MANAGEMENT` for drivers that will need to control the cache later.
If removed "cache", there will no cache driver could be used in later. In fact, excluding L2 cache operations, everything else is compatible.

jimmyzhe replied:
Sorry, I mean the following code enables Andes Coherence Manager hardware feature.
```
	li t0, (1 << 19) | (1 << 13) | (1 << 10) | (1 << 1)
	csrs NDS_MCACHE_CTL, t0

	/* Check if CPU support CM or not. */
	csrr t0, NDS_MCACHE_CTL
	li t1, (1 << 19)
	and t0, t0, t1
	beqz t0, cache_enable_finish

	/* If CPU support CM, check if CM is enabled. */
	li t1, (1 << 20)
check_cm_enabled:
	csrr t0, NDS_MCACHE_CTL
	and t0, t0, t1
	beqz t0, check_cm_enabled

cache_enable_finish:
```
If ET171 doesn't support L2 cache (Andes Coherence Manager depends on this), here could be simplified to 
```
	li t0, (1 << 13) | (1 << 10) | (1 << 1)
	csrs NDS_MCACHE_CTL, t0
```

jimmyzhe commented on soc/egis/et171/start.S:
```diff
@@ -0,0 +1,75 @@
+	/* Disable linker relaxation before GP register initialization. */
+	.option push
+	.option norelax
+
```
I think RISC-V assembler should support `.align 2` ([riscv-asm-manual/src/asm-manual.adoc](https://github.com/riscv-non-isa/riscv-asm-manual/blob/main/src/asm-manual.adoc#pseudo-ops)). I would suggest using it like
```
SECTION_FUNC(init, entry)
	/* Disable linker relaxation before GP register initialization. */
	.option push
	.option norelax	# prevent RISC-V linker relaxation
	.option norvc	# prevent "j header_end" optimized to compressed instruction

	j header_end
	.align 2		# padding for "j header_end" optimized to Andes EXECIT 16 bit insturciton
	.word 0x53494745 /* "EGIS" */
	...
```

Add `nop` after `j header_end` also works, just note that this depends on both instructions must be optimized to compressed instruction.

jimmyzhe replied:
Thanks for your feedback!
I can reproduce this with "ZEPYHR_TOOLCHAIN_VARIANT=llvm".

I think separating start.S to build with `-mno-execit` could be an option, so that no need to rely on compressed instruction.
```
        /* Disable linker relaxation before GP register initialization. */
        .option push
        .option norelax
+       .option norvc
 
        j header_end
-       nop /* @TODO: use to align. ".align" 2 doesn't work */
        .word 0x53494745 /* "EGIS" */
        ...
```

```
 zephyr_sources(
   soc.c
-  start.S
   soc_irq.S
 )
 
+zephyr_library()
+zephyr_library_sources(start.S)
 ...

if(CONFIG_SOC_EGIS_EXECIT)
   zephyr_cc_option(-mexecit)
+  zephyr_library_compile_options(-mno-execit)
   zephyr_ld_options(-Wl,--mexecit)
 endif()
...
```

jimmyzhe reviewed (commented):
#90307 has already merged, is that possible to add spi node in this PR? or there is additional plan for ET171 SPI driver (and flash driver?).

fabiobaltieri commented:
Hey folks, had a chat with my colleagues working on this, my understanding is that while this is not off the shelf, it is a developer kit provided to end customers, not unlike many other boards in the tree (some the ARM ones I thinks, the Nordic PDK that have been removed recently, actually those were temporary this one is not from what I understand). I think that there's a point in having it upstream as the scope of that board is not to have it paired with a specific application, but to have downstream customers developing on it.

In addition to that, it's my understanding that we are going to introduce a Google board with it down the road as well, so the soc and drivers are very much needed. Obviously everything could go on a module but I think there's good reasons to have these upstream.

@kartben @nordicjm what's missing from the board documentation file beside the picture? I can't quite find the requirements for upstream boards.

kartben commented on boards/egis/egis_et171/doc/index.rst:
```diff
@@ -0,0 +1,83 @@
+========
+
+Since this is a B2B prototype, there is no publicly available
+flashing tool. If you have any development needs, please contact us.
```
"us" is the Zephyr Project so that wording certainly needs to be changed.

jackylee-go replied:
Thank you for your reminder, I will clearly guide to Egis.

jackylee-go commented on soc/egis/et171/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+
+# Note: AndeStar V5 DSP needs custom Andes V5 toolchain
+if(CONFIG_SOC_EGIS_ET171_HWDSP)
+  zephyr_cc_option(-mext-dsp)
```
As I known this argument `-mext-dsp` is only for Andes ToolChain. ( same as `-mexecit` )
@jimmyzhe mentioned that, they have a [PR](https://github.com/zephyrproject-rtos/zephyr/pull/93697) for this kind of modification, but it is still under review.
Unless their PR is merged, I assume any changes I make for this will conflict with theirs.
So I thought I would remove it first, since those argument would be no effect on the standard compiler.

jackylee-go commented on soc/egis/et171/Kconfig:
```diff
@@ -0,0 +1,55 @@
+	select HAS_FLASH_LOAD_OFFSET
+	select SOC_EARLY_INIT_HOOK
+
+config SOC_EGIS_ET171_HWDSP
```
For the same reason as [above](https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2351067608), I would remove it first.

jimmyzhe commented on soc/egis/et171/linker.ld:
Thanks for reporting this, I need some time to dig out this problem.
```
/zephyrproject/zephyr/arch/riscv/core/pmp.c:358:(.text.z_riscv_pmp_init+0x4): relocation truncated to fit: R_RISCV_GPREL_I against symbol `__rom_region_size' defined in *ABS* section in zephyr/zephyr_pre0.elf
```

I think the root cause is that the ET171/AE350 XIP ROM and RAM layout may cause the **offset between `__rom_region_size` and `__global_pointer` exceed the GP relaxation range (+/-0x800) after linker relaxation.**

-----
Here is my experiment:
Modify zephyr/soc/egis/et171/linker.ld to use RISC-V generic script.
```
#include <zephyr/arch/riscv/common/linker.ld>

SECTIONS
{
        SECTION_PROLOGUE(.egis_padding_section,,)
        {
        . = ALIGN(32);
        } GROUP_LINK_IN(ROMABLE_REGION)
        __egis_binary_end = LOADADDR(.egis_padding_section) + SIZEOF(.egis_padding_section);
        __egis_binary_size = __egis_binary_end - __rom_region_start;
}
```

Run `./scripts/twister -c -v -p egis_et171 -T samples/subsys/zbus/benchmark`

- Before/No linker relaxation (-mno-relax):
<img width="4247" height="1741" alt="擷取220" src="https://github.com/user-attachments/assets/973ada56-4b99-450c-bb78-f905e8bad2c6" />

- After/With linker relaxation (fail with R_RISCV_GPREL_I):
<img width="4247" height="1756" alt="擷取221" src="https://github.com/user-attachments/assets/b5bc05f1-7b2d-4b91-8826-f85d37d3c72e" />

-----
Although I used to add `rom_mpu_padding` for XIP [ae350/linker.ld](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/andestech/ae350/linker.ld#L440-L443) and seems work (I added this for PMP NAPOT padding originally).
- With linker relaxation and add `rom_mpu_padding`
(pass because the offset is fixed no matter linker relaxation)
<img width="4247" height="1873" alt="擷取222" src="https://github.com/user-attachments/assets/f75cb2da-5bb2-4cd5-aecd-124443e1a91f" />

This problem still happens when `CONFIG_PMP_POWER_OF_TWO_ALIGNMENT=n`.

-----
I think this problem also occur on the AE350, because RAM based at 0x00000000 (place `__global_pointer` too close to `__rom_region_size`).

I plan to fix this in zephyr/soc/andestech/ae350/CMakeLists.txt like:
```
if(CONFIG_XIP AND CONFIG_RISCV_GENERIC_TOOLCHAIN)
  zephyr_ld_options(-mno-relax)
endif()
```
The reason for add CONFIG_RISCV_GENERIC_TOOLCHAIN is that the Andes toolchain
and extended instructions can increase the GP relaxation range to +/-0x20000.

I am wandering if this approach acceptable for ET171.

jackylee-go replied:
Thank you for your detailed explanation.
I've tested your solution, and it works.
Although I think it's the best solution, as it helps keep my ld file cleaner.
But another issue is that, as @nordicjm  mentioned (https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2350983723), I believe the compiler argument"-mno-relax" also needs to be moved to `cmake`.
I think this would complicate the PR, especially since it would conflict with [yours](https://github.com/zephyrproject-rtos/zephyr/pull/93697).
If you don't mind, I'd like to keep it as is until your PR is merged, and then I'll make the changes.
Alternatively, I can open a PR to address this issue after our PR is merged.

jimmyzhe replied:
yeah, we can remove Andes toolchain option and CSR first and extend these in the future.

However, the `-mno-relax` option here is more like a soc specified usage, since the AE350/ET171 ROM/RAM layout cause this problem.

I would suggest not to use [ae350/linker.ld](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/andestech/ae350/linker.ld#L440-L443)  `__rom_region_size` to workaround this, because this still happens when `CONFIG_PMP_POWER_OF_TWO_ALIGNMENT=n`.

jackylee-go replied:
If there is no need to modify the cmake content.
We would like to adopt your suggestion because I think it is the best solution for us.

Thank a lot for your kindly help.

fabiobaltieri commented on boards/egis/egis_et171/doc/index.rst:
```diff
@@ -0,0 +1,83 @@
+========
+
+Since this is a B2B prototype, there is no publicly available
+flashing tool. If you have any development needs, please contact us.
```
sure that's fine but then this blocks having a proper integration with the integrated flashing infrastructure for the sake of hiding basically the name of the prioprietary flashing binary and flags and nothing else, basically I'm saying you could do something like this: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/beagle/beagleconnect_freedom/board.cmake, just say that the tool is part of the super secret but hey if you have it you can now run `west flash` and it'll work.

Anyway, up to you, I understand that there's some bureocratic hassle related to this stuff and there's other boards in tree in a similar situation.

jimmyzhe reviewed (dismissed):
LGTM, I will update ET171 with Andes custom CSR in #93697 after this PR merged.

ThomasChanAtEgisTec commented:
The PR has already got two approvals but it is still not merged.
Does anyone know the reason?

jackylee-go commented:
@jimmyzhe @fabiobaltieri
I have added the `board.cmake` file for the [mentaion](https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2352733333).
Please review it again, thanks.

kartben commented:
> The PR has already got two approvals but it is still not merged.
> Does anyone know the reason?

And it has pending request for changes for which you need to leave time to the people who made them to revisit and approve

fabiobaltieri commented:
> @jimmyzhe @fabiobaltieri I have added the `board.cmake` file for the [mentaion](https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2352733333). Please review it again, thanks.

Great stuff thanks a lot.

> The PR has already got two approvals but it is still not merged.
> Does anyone know the reason?

Yeah as Ben pointed out, it'll get merged once it gets all blockers resolved and two approvals including the maintainer, no need to worry about it just handles the review comments someone else will take care of the rest.

ThomasChanAtEgisTec commented:
Hi, does this mean that the PR needs that the requested changes are resolve? I think Jacky has made changes to fulfill the requested changes from @kartben and @nordicjm.

nordicjm reviewed (commented):
if you want people to re-review, click the refresh button next to names under the review list

jackylee-go commented on boards/egis/egis_et171/board.cmake:
```diff
@@ -0,0 +1,25 @@
+# As per the NDA you signed with us, you will receive the firmware
+# flashing tools and related documents.
+
+# You must set the private key of the secure boot upgrade from env.
```
The following excerpt is from [bin_blobs.html#limited-scope](https://Docs.Zephyrproject.Org/latest/contribute/bin_blobs.Html#limited-scope)
```
The project maintainers define an open source test suite that an upstream target must be able to pass using only open source software included in the mainline distribution and its modules. The open source test suite currently consists of:

samples/philosophers

tests/kernel
```
I had tested the `samples/philosophers` project on our board, and it worked fine.
If I understand correctly, the CI script has already tested `tests/kernel`, and the test [passed](https://github.com/zephyrproject-rtos/zephyr/actions/runs/17788023980/job/50559360325?pr=91940).

If there is any other info needed, please let me know.

jackylee-go replied:
I also tested the `tests/kernel` locally, and it appears to have passed the tests.
```
(.venv) jacky@TPE-JACKY-LEE1:~/zephyrproject_upstream/zephyr$ west twister -T tests/kernel -p egis_et171
INFO    - Using Ninja..
INFO    - Zephyr version: aa87d80545a1
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/jacky/zephyrproject_upstream/zephyr/twister-out/testplan.json
INFO    - JOBS: 12
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  127/ 127  100%  built (not run):  114, filtered:   66, failed:    0, error:    0
INFO    - 179 test scenarios (180 configurations) selected, 66 configurations filtered (53 by static filter, 13 at runtime).
INFO    - 0 of 114 executed test configurations passed (0.00%), 114 built (not run), 0 failed, 0 errored, with no warnings in 734.14 seconds.
INFO    - 0 of 0 executed test cases passed (0.00%) on 0 out of total 1170 platforms (0.00%).
INFO    - 1539 selected test cases not executed: 1539 not run (built only).
INFO    - 0 test configurations executed on platforms, 114 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/jacky/zephyrproject_upstream/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/jacky/zephyrproject_upstream/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/jacky/zephyrproject_upstream/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

kartben replied:
> so what happens without any of this? How does someone flash this board using upstream or use it with twister?

FWIW we have some boards with no support for flashing already (but maybe we shouldn't?)
https://docs.zephyrproject.org/latest/boards/beagle/beagley_ai/doc/index.html

jackylee-go replied:
Actually, we didn't plan to provide firmware update functionality to end users. However, based on @fabiobaltieri  's [suggestion](https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2352733333), we added a note to the `board.cmake`, informing users that technical support from Egis is required to use this feature.

Please let me know if you have any other concerns.

jimmyzhe replied:
Hi @jackylee-go,

I noticed that egis_et171 can be debugged with the Andes ICE debugger, and it can be built with `CONFIG_XIP=n` (run-from-RAM).

Is this a valid workflow on ET171 for using the Andes ICE debugger and default ICEman with `CONFIG_XIP=n` (load code by GDB)?

If yes, would you mind to add a flashing section for the `CONFIG_XIP=n` workflow, similar to the [ADP-XC7K AE350 flashing guide](https://docs.zephyrproject.org/latest/boards/andestech/adp_xc7k_ae350/doc/index.html#flashing)? Some board also support load code by GDB only, like [HiFive Unmatched](https://docs.zephyrproject.org/latest/boards/sifive/hifive_unmatched/doc/index.html#flashing).

The Andes ICEman packages is available here: [Andes-Development-Kit releases](https://github.com/andestech/Andes-Development-Kit/releases)

nordicjm replied:
This board.cmake file requires external input is what I mean, you can't just use it in CI for example and it uses environmental values, ping @tejlmand to see his views on this

jackylee-go commented on boards/egis/egis_et171/board.cmake:
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+# To get more info about flashing, please contect https://www.egistec.com
+# As per the NDA you signed with us, you will receive the firmware
+# flashing tools and related documents.
```
Thank you for your reminder.
I'll revise the wording.

jackylee-go commented:
> > @kartben @nordicjm hey can you take another look when you have a chance? I think all comments have been addressed
> 
> @tejlmand needs to review [9ec8189#r2355220428](https://github.com/zephyrproject-rtos/zephyr/commit/9ec818949829ca41fe681f089ebf50b8adcc1b66#r2355220428) afaik there should be no linker flag usage like this by SoCs And there is the query on the part above, if this can't be used with zephyr as it is and non-NDA tools to run hello world on target, what use is it? How does that meet the minimum requirements for a board?

As you mentioned, I have already tagged @tejlmand and explained the reason.

Regarding the "-mno-relax" parameter, I think @jimmyzhe has already explained the specific reasons and suggestions in detail in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/91940#discussion_r2351501547).

However, I'm not sure if there's a more efficient way to get his review.
If @tejlmand provides any feedback, I will follow up asap.

jimmyzhe commented:
> @tejlmand needs to review [9ec8189#r2355220428](https://github.com/zephyrproject-rtos/zephyr/commit/9ec818949829ca41fe681f089ebf50b8adcc1b66#r2355220428) afaik there should be no linker flag usage like this by SoCs 

yeah..., it is weird to add the `-mno-relax` option in SoC.

I think the problem is RISC-V linker's limitation, the absolute symbol `__rom_region_size` may cross GP relaxation range in the AE350/ET171 XIP linker script.

I have two possible approaches to fix this ET171 issue
1. Make `config LINKER_USE_RELAX` default disable in zephyr/soc/egis/et171/Kconfig.defconfig.et171.
This is effectively the same as `-mno-relax`, and users can enable it if needed.
```
config LINKER_USE_RELAX
      default n if XIP
```
2. Overwrite the `__rom_region_size` value in zephyr/soc/egis/et171/linker.ld.
define this symbol as a constant that linker probably not relax it. Just note that the `__rom_region_size` no longer represent the actual ROM usage anymore. 
```
#include <zephyr/arch/riscv/common/linker.ld>

SECTIONS
{
...
#ifdef CONFIG_XIP
       __rom_region_size = ROM_SIZE;
#endif
 }
```

I think approach (1) would be better, since it lets users choose whether to enable linker relaxation and avoids changing symbol's meaning or generic code.

Hi all, if these approaches acceptable, I also plan to fix this in AE350, thanks.

mbolivar reviewed (commented):
no objections to the dts vendor-prefixes.txt change

jackylee-go commented:
The [twister-build](https://github.com/zephyrproject-rtos/zephyr/actions/runs/18123163841/job/51681180874?pr=91940) test failed during the CI process.
It seems the module "`jsonschema`" is not installed in the runner env.
I don't have permission to rerun the test without commit anything.
Could a maintainer please help resolve this issue?
```
Run export ZEPHYR_BASE=${PWD}
Traceback (most recent call last):
  File "/home/runner/work/zephyr/zephyr/zephyr/./scripts/ci/test_plan.py", line 38, in <module>
    import list_boards
  File "/home/runner/work/zephyr/zephyr/zephyr/scripts/list_boards.py", line 14, in <module>
    import jsonschema
ModuleNotFoundError: No module named 'jsonschema'
Error: Process completed with exit code 1.
```

jimmyzhe commented:
It looks like the current codebase always hits this problem, could you rebase to main branch and try again?

Commit:
soc: Add Egis et171

This is a SOC based on AE350. In addition to the core, some
modifications have been made to the peripheral functions,
including the integration of built-in USB.

Signed-off-by: Jacky Lee <jacky.lee@egistec.com>

Commit:
driver: Reuse andes_cache driver for compatible MCUs

Fixed a wrong function declare of 'nds_l2_cache_init'.
Replace `SOC_FAMILY_ANDES_V5` with `DT_HAS_ANDESTECH_ANDESCORE_V5_ENABLED`
condition to enable it to be used by compatible MCUs

Signed-off-by: Jacky Lee <jacky.lee@egistec.com>

jimmyzhe commented on boards/egis/egis_et171/egis_et171_defconfig:
```diff
@@ -0,0 +1,8 @@
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# Serial driver options
+CONFIG_SERIAL=y
+
+# GPIO driver options
```
Sorry I saw this late. If this is from adp_xc7k_ae350_defconfig, could you also clean up it like [boards/andestech/adp_xc7k_ae350/adp_xc7k_ae350_defconfig](https://github.com/zephyrproject-rtos/zephyr/pull/93697/files#diff-82b69dd307d16bda801c816e44ab8348041157e32f7e8017a0d09e994f976e8e).

I will remove this in adp_xc7k_ae350 in other PR.
Otherwise, LGTM. Thanks.

jackylee-go replied:
Okay, no problem.
I will delete those unnecessary comment lines.

Commit:
board: Add 'egis_et171' dev board

This is a development board for Egis et171 SOC.
It has basic peripherals such as GPIO, I2C, SPI,.. etc
and built-in USB.

Signed-off-by: Jacky Lee <jacky.lee@egistec.com>

jackylee-go commented:
@kartben, @nordicjm, your change requests are still marked.
Could you kindly confirm if there's anything that still needs to be changed?

kartben reviewed (approved):
Fixing my comments as a follow-up would be fine. Thanks for your patience!

jackylee-go commented on boards/egis/egis_et171/doc/index.rst:
```diff
@@ -0,0 +1,120 @@
+Connections and IOs
+===================
+
+The et171 platform has 1 GPIO controller. It providing 17 bits of IO.
```
Thank you for your careful review. I will correct these errors in future updates.
